OS-dependent symlink generation during installation

ideaconnect opened the pull request:
Fails on Windows with error:
```
yarn run v1.22.0
$ gulp build
[02:48:32] Requiring external module babel-register
[02:48:34] Using gulpfile D:\dev\bug\DHTest\tests\Application\gulpfile.babel.js
[02:48:34] Starting 'build'...
[02:48:34] Starting 'buildAdmin'...
[02:48:34] Starting 'buildShop'...
[gulp-chug] Gulpfile, gulpfile.babel.js, contents is empty. Reading directly from disk...
[02:48:34] 'buildAdmin' errored after 50 ms
[02:48:34] Error in plugin 'gulp-chug'
Message:
    Problem finding locally-installed `gulp` for gulpfile undefined. (Try running `npm install gulp` from <path>\DHTest\vendor\sylius\sylius\src\Sylius\Bundle\AdminBundle to install a local gulp for said gulpfile.)

Error: Cannot find module 'gulp' from '<path>\DHTest\vendor\sylius\sylius\src\Sylius\Bundle\AdminBundle'
Details:
    domainEmitter: [object Object]
    domain: [object Object]
    domainThrown: false
[02:48:34] 'build' errored after 60 ms
[02:48:34] The following tasks did not complete: buildShop
[02:48:34] Did you forget to signal async completion?
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```


Steps to reproduce:
1) `composer create-project sylius/plugin-skeleton ProjectName`
2) go to the installation folder
3) execute: `(cd tests/Application && yarn install)`
4) go to the installation folder, execute: `(cd tests/Application && yarn build)`

Issue suggested by the message:
```
Problem finding locally-installed `gulp` for gulpfile undefined. (Try running `npm install gulp`
```
is actually not the problem.

Symlink in a format not accepted by Windows is:
https://github.com/Sylius/PluginSkeleton/blob/master/node_modules

Solution:
use a way to generate the symlink for the local platform.

pamil commented on composer.json:
```diff
@@ -52,5 +52,13 @@
     "config": {
         "sort-packages": true
+    },
+    "scripts": {
```
Could we also add it to `post-create-project-cmd`?

vvasiloi commented:
@ideaconnect did you try PHP's built-in, cross-platform functions: [symlink](https://www.php.net/manual/en/function.symlink.php) and [link](https://www.php.net/manual/en/function.link.php)
Also, maybe first check if the link doesn't exist.

ideaconnect commented:
I have rebased on `master` yet from what I can see there is an error with services when running in `test` environment and is related to the latest update in Sylius itself. Fixed by update of the bundles config:
```
ApiPlatform\Core\Bridge\Symfony\Bundle\ApiPlatformBundle::class => ['all' => true],
```

Later an error:
```
Symfony\Component\Routing\Exception\RouteNotFoundException: Unable to generate a URL for the named route "api_doc" as such route does not exist. in vendor/symfony/routing/Generator/CompiledUrlGenerator.php:50
``` 
started to occur and it seems that the reason was lack of configuration related to api platform from latest Sylius/Sylius.

I have added minimum of compatibility updates for 1.8-dev for the existing tests to pass, yet probably you have your own procedure.

@pamil @vvasiloi 

On Windows: 
* `link` does not support folders, gives permission denied even with admin privileges
* `symlink` to folders works only with admin privileges and requires target path to be already present on Windows - in: `./win32/ioutil.c` of PHP's code:
```
        if ((attr = GetFileAttributesW(target)) == INVALID_FILE_ATTRIBUTES) {
                SET_ERRNO_FROM_WIN32_CODE(GetLastError());
                return -1;
        }
```

Technically the problem of path being present is not our case here as the empty `node_modules` folder already exists in `Application` yet changes made now are prepared for a scenario when such folder does not yet exist and is created after `yarn` execution.

Functions respectively rely on:
`php_win32_ioutil_symlink` and `php_win32_ioutil_symlink` calls (as defined in `./Zend/zend_virtual_cwd.h` of the php src)
which rely on:
`php_win32_ioutil_symlink_w` and `php_win32_ioutil_link_w`
which translate into:
* `php_win32_ioutil_symlink_w` => `CreateSymbolicLinkW(link, target, (attr & FILE_ATTRIBUTE_DIRECTORY ? 1 : 0));`
* `php_win32_ioutil_link_w` => `CreateHardLinkW(link, target, NULL);`


As microsoft says in their docs:
https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createhardlinkw
`CreateHardLinkW` works only with files.

and `CreateSymbolicLinkW` can be used only without admin privileges if another flag is set (not set in the code of php) and system is in the `developer mode`.
https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinkw

Therefore I suggest to always run the function `symlink` yet on windows, if it fails fallback to a Junction. 

Since Junctions support already relative paths i guess there is no need for my additional approach of creating Symbolic Links (that is a different term on Windows).

In case a file already exists and is a link or directory I do not attempt to re-create it.
